src/main/java/com/billyyccc/mssqlclient/impl/codec/PreLoginCommandCodec.java

```diff
@@ -0,0 +1,109 @@
+package com.billyyccc.mssqlclient.impl.codec;
+
+import com.billyyccc.mssqlclient.impl.command.PreLoginCommand;
+import com.billyyccc.mssqlclient.impl.protocol.MessageStatus;
+import com.billyyccc.mssqlclient.impl.protocol.MessageType;
+import com.billyyccc.mssqlclient.impl.protocol.TdsMessage;
+import com.billyyccc.mssqlclient.impl.protocol.client.prelogin.EncryptionOptionToken;
+import com.billyyccc.mssqlclient.impl.protocol.client.prelogin.OptionToken;
+import com.billyyccc.mssqlclient.impl.protocol.client.prelogin.VersionOptionToken;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.vertx.sqlclient.impl.command.CommandResponse;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+class PreLoginCommandCodec extends MSSQLCommandCodec<Void, PreLoginCommand> {
+
+  PreLoginCommandCodec(PreLoginCommand cmd) {
+    super(cmd);
+  }
+
+  @Override
+  void encode(TdsMessageEncoder encoder) {
+    super.encode(encoder);
+    sendPreLoginMessage();
+  }
+
+  @Override
+  void decodeMessage(TdsMessage message, TdsMessageEncoder encoder) {
+    // nothing to do for now?
+    completionHandler.handle(CommandResponse.success(null));
+  }
+
+  private void sendPreLoginMessage() {
+    ChannelHandlerContext chctx = encoder.chctx;
+
+    ByteBuf packet = chctx.alloc().ioBuffer();
+
+    // packet header
+    packet.writeByte(MessageType.PRE_LOGIN.value());
+    packet.writeByte(MessageStatus.NORMAL.value() | MessageStatus.END_OF_MESSAGE.value());
+    int packetLenIdx = packet.writerIndex();
+    packet.writeShort(0); // set length later
+    packet.writeShort(0x00);
+    packet.writeByte(0x00); // FIXME packet ID
+    packet.writeByte(0x00);
+
+    // packet data
+    int packetDataStartIdx = packet.writerIndex();
+
+    List<OptionToken> optionTokens = cmd.optionTokens();
+
+    int payloadStartIdx = packet.writerIndex();
+
+    int totalLengthOfOptionsData = 0;
+
+      /*
+        we use a map to store offset and length related information:
+        payload ByteBuf writer position for Option offset ->  Option length
+        then we set offset by calculating ByteBuf writer indexes diff later
+       */
+    Map<Integer, Integer> offsetLengthInfo = new HashMap<>();
```

Comment:
this does not seem necessary too

Reply:
updated with https://github.com/BillyYccc/incubator-vertx-mssql-client/pull/3/commits/42efaec042cd1db791c609980a9b96c1444512c0.